File: tb3_square_driver.cpp
#include "rclcpp/rclcpp.hpp"
#include "tb3_motion_functions.hpp"

// Change these numbers first.
const double ForwardSpeed = 0.15;
const double TurnSpeed = 0.50;
const double ForwardTime = 4.0;
const double TurnTime = 3.2;
const double TimerPeriod = 0.1;

class TB3SquareDriver : public rclcpp::Node
{
public:
  TB3SquareDriver() : Node("tb3_square_driver")
  {
    cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 10);

    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(TimerPeriod),
      std::bind(&TB3SquareDriver::timer_callback, this));

    is_turning_ = false;
    elapsed_time_ = 0.0;

    RCLCPP_INFO(this->get_logger(), "tb3_square_driver started");
    RCLCPP_INFO(this->get_logger(), "robot moves forward, then turns left");
  }

private:
  void timer_callback()
  {
    if (is_turning_) {
      cmd_vel_pub_->publish(make_turn_left_cmd(TurnSpeed));
    } else {
      cmd_vel_pub_->publish(make_forward_cmd(ForwardSpeed));
    }

    elapsed_time_ += TimerPeriod;

    if (!is_turning_ && elapsed_time_ >= ForwardTime) {
      is_turning_ = true;
      elapsed_time_ = 0.0;
      RCLCPP_INFO(this->get_logger(), "turn");
    } else if (is_turning_ && elapsed_time_ >= TurnTime) {
      is_turning_ = false;
      elapsed_time_ = 0.0;
      RCLCPP_INFO(this->get_logger(), "forward");
    }
  }

  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  bool is_turning_;
  double elapsed_time_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<TB3SquareDriver>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: tb3_square_driver.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist


class TB3SquareDriver(Node):
    def __init__(self) -> None:
        super().__init__("tb3_square_driver")

        self.publisher_ = self.create_publisher(Twist, "/cmd_vel", 10)
        self.timer = self.create_timer(0.1, self.timer_callback)

        # 状態管理
        self.mode = "forward"
        self.mode_time = 0.0

        # 調整しやすいパラメータ
        self.forward_speed = 0.15
        self.turn_speed = 0.5
        self.forward_duration = 4.0
        self.turn_duration = 3.2

        self.get_logger().info("TB3 square driver started.")

    def timer_callback(self) -> None:
        dt = 0.1
        self.mode_time += dt

        cmd = Twist()

        if self.mode == "forward":
            cmd.linear.x = self.forward_speed
            cmd.angular.z = 0.0
            if self.mode_time >= self.forward_duration:
                self.mode = "turn"
                self.mode_time = 0.0
                self.get_logger().info("Switch to TURN")

        elif self.mode == "turn":
            cmd.linear.x = 0.0
            cmd.angular.z = self.turn_speed
            if self.mode_time >= self.turn_duration:
                self.mode = "forward"
                self.mode_time = 0.0
                self.get_logger().info("Switch to FORWARD")

        self.publisher_.publish(cmd)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = TB3SquareDriver()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        # 終了時に停止命令
        stop_msg = Twist()
        node.publisher_.publish(stop_msg)
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()